packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,155 @@
+        const state = stateByRunId.get(runId);
+        if (state) {
+          states.push(state);
+        } else {
+          const runState = await this.getSingleState(runId, await this._accessToken());
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = BulkExtractor.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  private async fetchStates() {
+    this._reportStates = await this.getStates(this._reportIds);
+    this._timeFetched = new Date();
+  }
+
+  public getState(reportId: string) {
+    if ((new Date().getTime() - this._timeFetched.getTime()) > STATUS_CHECK_INTERVAL) {
+      this.fetchStates().catch((e) =>
+        /* eslint-disable no-console */
+        console.error(e)
+      );
+    }
+    return this._reportStates.get(reportId) ?? ExtractionStates.None;
+  }
+
+  public clearJob(reportId: string) {
+    this._reportRunIds.delete(reportId);
+  }
+
+  private static getFinalState(states: ExtractorState[]) {
+    if (states.includes(ExtractorState.Failed))
+      return ExtractionStates.Failed;
+
+    if (states.includes(ExtractorState.Queued))
+      return ExtractionStates.Queued;
+
+    if (states.includes(ExtractorState.Running))
+      return ExtractionStates.Running;
+
+    if (states.includes(ExtractorState.Succeeded))
+      return ExtractionStates.Succeeded;
+
+    return ExtractionStates.Failed;
+  }
+
+  private async getSingleState(runId: string, accessToken: string): Promise<ExtractorState> {
+    try {
+      const response = await this._extractionClientApi.getExtractionStatus(accessToken, runId);
+      return response.state;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return ExtractorState.Failed;
+  }
+
+  public async startJobs(reportIds: string[]) {
+    const reportIModelIds = new Map<string, string[]>();
+    for (const reportId of reportIds) {
+      const reportIModels = await this.fetchReportIModels(reportId);
+      reportIModelIds.set(reportId, reportIModels);
+      this._reportStates.set(reportId, ExtractionStates.Starting);
+    }
+    const iModels = new Set(Array.from(reportIModelIds.values()).flat());
+    const extractionMap =
+      Array.from(iModels).map(async (iModel): Promise<[string, string | undefined]> => {
+        const run = await this.runExtraction(iModel);
+        return [iModel, run?.id];
+      });
+
+    const extractionByIModel = new Map<string, string | undefined>(await Promise.all(extractionMap));
+    reportIds.forEach((reportId) => {
+      const reportIModels = reportIModelIds.get(reportId)!;
+      const runs: string[] = [];
+      reportIModels.forEach((iModelId) => {
+        const runId = extractionByIModel.get(iModelId);
+        if (runId)
+          runs.push(runId);
+      });
+      this._reportRunIds.set(reportId, runs);
+    });
+  }
+
+  private async runExtraction(iModelId: string) {
```

Comment:
Add return types to all methods. That should make it easier to understand what is being returned. I think returning the `id` (`string | undefined`) from `runExtraction` would be enough.

Reply:
Return types added to all methods.

Reply:
Still not all methods have return types. You missed some private methods.

Reply:
Missing types added

Reply:
Confirmed.